[Agents] Adding agent websocket for web and mobile clients (#172)
cartesia-ai/docs: fern/agents/integrations/web-calls.mdx

# Web Calls

> Some Cartesia customers might be interested in integrating their Line agents with their website, rather than a telephony provider. For these folks, the Agents WebSocket provides real-time, bidirectional communication with voice agents through a streaming interface. This allows you to send audio data and receive agent responses in real-time for any client.

Reviewer: ```suggestion
The Agents WebSocket provides real-time, bidirectional communication between web clients and Cartesia voice agents. It enables streaming audio input and real-time agent responses for browser-based or custom applications.
```
public-facing language

```mermaid
graph LR
    A1[Web Client] <-->|General Websocket Events| B[Cartesia API]

>     B <-->|Agent Events| C[Inferno Agents]

Reviewer: ```suggestion
    B <-->|Agent Events| C[Cartesia Agents]
```

(important)

```

## Connection

Connect to the WebSocket endpoint:

```
wss://api.cartesia.ai/agents/stream/{agent_id}
```

**Headers:**
- `Authorization: Bearer {your_api_key}`

> - `Cartesia-Version: 2025-04-16`

Reviewer: we should use a table here

## Protocol Overview

The WebSocket protocol uses JSON messages for control and base64-encoded audio for media. The connection follows this flow:

1. Client sends `start` event to initialize the stream
2. Server responds with `ack` event confirming configuration

> 3. Bidirectional exchange of events

Reviewer: After acknowledgment, the client and server exchange streaming audio and control events until either side closes the connection, or the inactivity timeout is fired.

  "config": {
    "input_format": "pcm_44100"
  },
  "metadata": {
    "to": "user@example.com",
    "from": "+1234567890"
  }
}
```

**Fields:**
- `stream_id` (optional): Stream identifier. If not provided, server generates one
- `config.input_format`: Audio format for client audio input (`mulaw_8000`, `pcm_16000`, `pcm_24000`, `pcm_44100`)
- `metadata` (optional): Custom metadata object. These will be passed through to the user code, but there are some special fields you can use as well:
  - `to` (optional): Destination identifier for call routing (defaults to agent ID)
  - `from` (optional): Source identifier for the call (defaults to "websocket")

### Media Input Event

Audio data sent from the client to the server. `payload` audio data should be base64 encoded.

```json
{
  "event": "media_input",
  "stream_id": "example_id",
  "media": {
    "payload": "base64_encoded_audio_data"
  }
}
```

**Fields:**
- `stream_id`: Stream identifier from the ack response
- `media.payload`: Base64-encoded audio data in the format specified in the start event

### DTMF Event

Sends DTMF (dual-tone multi-frequency) tones.

```json
{
  "event": "dtmf",
  "stream_id": "example_id",

>   "dtmf": "1"

Author: synced irl - this is predominantly used in the telephony usecase, so we'll keep it dtmf

# Web Calls

The Agents WebSocket provides real-time, bidirectional communication between web clients and Cartesia voice agents. It enables streaming audio input and real-time agent responses for browser-based or custom applications.

```mermaid
graph LR
    A1[Web Client] <-->|General Websocket Events| B[Cartesia API]
    B <-->|Agent Events| C[Cartesia Agents]
```

## Connection

Connect to the WebSocket endpoint:

```
wss://api.cartesia.ai/agents/stream/{agent_id}
```

**Headers:**

| Header | Value |
|--------|-------|

> | `Authorization` | `Bearer {your_api_key}` |

Reviewer 2: Token, not API key.

Author: Oh i thought it could be either - i'll update tho.

  }
}
```

**Fields:**
- `stream_id` (optional): Stream identifier. If not provided, server generates one
- `config.input_format`: Audio format for client audio input (`mulaw_8000`, `pcm_16000`, `pcm_24000`, `pcm_44100`)
- `metadata` (optional): Custom metadata object. These will be passed through to the user code, but there are some special fields you can use as well:
  - `to` (optional): Destination identifier for call routing (defaults to agent ID)
  - `from` (optional): Source identifier for the call (defaults to "websocket")

### Media Input Event

Audio data sent from the client to the server. `payload` audio data should be base64 encoded.

```json
{
  "event": "media_input",
  "stream_id": "unique_id",
  "media": {
    "payload": "base64_encoded_audio_data"
  }
}
```

**Fields:**
- `stream_id`: Unique identifier for the Stream from the ack response
- `media.payload`: Base64-encoded audio data in the format specified in the start event

### DTMF Event

Sends DTMF (dual-tone multi-frequency) tones.

```json
{
  "event": "dtmf",
  "stream_id": "example_id",
  "dtmf": "1"
}
```

**Fields:**
- `stream_id`: Stream identifier
- `dtmf`: DTMF digit (0-9, *, #)

### Custom Event

> Sends custom metadata to the agent.

Reviewer 2: Probably should explain how this shows up to agent code?

Author: Heh it does not yet show up, we have a ticket for this outstanding

## Server events

### Ack Event

Server acknowledgment of the start event, confirming stream configuration.
If `stream_id` wasn't provided in the initial `start` event, this is where the user can obtain the server generated `stream_id`.

```json
{
  "event": "ack",
  "stream_id": "example_id",
  "config": {
    "input_format": "pcm_44100"
  }
}
```

### Media Output Event

Server sends agent audio response. `payload` is base 64 encoded audio data.

```json
{
  "event": "media_output",
  "stream_id": "example_id",
  "media": {
    "payload": "base64_encoded_audio_data"
  }
}
```

### Clear Event

Indicates the agent wants to clear/interrupt the current audio stream.

```json
{
  "event": "clear",
  "stream_id": "example_id"
}
```

### DTMF Event

Server sends DTMF tones from the agent.

```json
{
  "event": "dtmf",
  "stream_id": "example_id",
  "dtmf": "5"
}
```

### Custom Event

Server sends custom metadata from the agent.

```json
{
  "event": "custom",
  "stream_id": "example_id",
  "metadata": {
    "agent_state": "processing",
    "confidence": 0.95,
    "custom_data": "value"
  }
}
```

## Connection Management

### Inactivity Timeout

> The server automatically closes idle WebSocket connections after **30 seconds** of inactivity. Activity is defined as receiving any message from the client, including:

Reviewer: ```suggestion
The server closes idle WebSocket connections after **30 seconds** without client activity. Any client message counts as activity, including:
```

- Application messages (media_input, dtmf, custom events)
- Standard WebSocket ping frames
- Any other valid WebSocket message

When the timeout occurs, the connection is closed with:
- **Code:** 1000 (Normal Closure)
- **Reason:** `"connection idle timeout"`

### Ping/Pong Keepalive

> To prevent inactivity timeouts during periods of silence, use standard WebSocket ping frames for periodic keepalive:

Reviewer: ```suggestion
Send periodic WebSocket ping frames to keep the connection alive:
```

Reviewer: periods of silence is misleading

Author: @sauhardjain i think the issue is we don't want folks to keep this alive indefinitely without reason. Although we'll get paid (lol) they'll probably get pissed

```python
# Client sends ping to reset inactivity timer
pong_waiter = await websocket.ping()
latency = await pong_waiter
```

```javascript
// JavaScript example
setInterval(() => {
  if (websocket.readyState === WebSocket.OPEN) {

>     websocket.ping();

Reviewer: `websocket.ping()` isn't supported on browsers, so we should make this example specific to Node.js

  }

> }, 20000); // Send ping every 20 seconds

Reviewer: ```suggestion
}, 20000); // every 20s to avoid 30s idle timeout
```

```

> 

Reviewer: Add browser-specific example:

```
// Browser example: send a custom keepalive event
setInterval(() => {
  if (websocket.readyState === WebSocket.OPEN) {
    websocket.send(JSON.stringify({
      event: "custom",
      stream_id: "unique_id",
      metadata: { type: "heartbeat" }
    }));
  }
}, 20000); // every 20s to avoid 30s idle timeout
```

The server automatically responds to ping frames with pong frames and resets the inactivity timer upon receiving any message.

### Connection Close

The connection can be closed by either the client or server using WebSocket close frames.

**Client-initiated close:**
```python
await websocket.close(code=1000, reason="session completed")
```

**Server-initiated close:**
When the agent ends the call, the server closes the connection with:
- **Code:** 1000 (Normal Closure)
- **Reason:** `"call ended by agent"` or `"call ended by agent, reason: {specific_reason}"` if additional context is available

## Best Practices

1. **Always send start event first** - The connection will be closed if any other event is sent before start

> 2. **Use appropriate audio formats** - Match your input format to your audio source capabilities. For telephony providers this is often `mulaw_8000`, while for web clients this will often be `pcm_44000`

Reviewer 2: IMO we should recommend 16k

Author: @noahlt the input gets resampled to 16k by default in our Pipecat pipeline since that's what our STT takes, independent of what the input Transport takes.

But right now input Transport also informs our output (we have a ticket to fit this)

> 1. **Always send start event first** - The connection will be closed if any other event is sent before start
> 2. **Use appropriate audio formats** - Match your input format to your audio source capabilities. For telephony providers this is often `mulaw_8000`, while for web clients this will often be `pcm_44000`
> 3. **Handle connection close gracefully** - Monitor close events and reasons for debugging
> 4. **Implement keepalive for calls with longer periods of silence** - Send WebSocket ping frames every 20-25 seconds to prevent the 30-second inactivity timeout during periods of silence
> 5. Send your own stream_id's for the best observability
> 6. Always handle timeout closures (`1000 / connection idle timeout`) by reconnecting and resending a `start` event.

Reviewer: ```suggestion
1. **Send `start` first** — The connection closes if any other event is sent before `start`.
1. **Choose the right audio format** — Match the format to your source: `mulaw_8000` for telephony, `pcm_44100` for web clients.
1. **Handle closes cleanly** — Always capture close codes and reasons for debugging and recovery.
1. **Keep the connection alive** — Send WebSocket ping frames every 20–25 seconds to avoid the 30-second inactivity timeout.
1. **Manage stream IDs** — Provide your own `stream_id` values to improve observability across systems.
1. **Recover from idle timeouts** — On `1000 / connection idle timeout`, reconnect and resend a `start` event.
```
